### PR TITLE
PPSD: add special handling and noise models for Infrasound

### PR DESCRIPTION
**Pull request by kamacpherson**
### What does this PR do?
This PR adds a "special_handling = 'infrasound'" option to the PPSD class. When invoked, this option removes the differentiation to acceleration step in the response removal from the PSDs (similar to the hydrophone case) and also adds global infrasound high and low noise models (Brown, 2012) for psd plotting.

### Why was it initiated?  Any relevant Issues?
I needed the ability to compute and visualize noise spectra for infrasound channels. I think it could be useful for others working in infrasound, particularly with the deployment of many BDF channels with the TA network.

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [x] If the PR is making changes to documentation, docs pages can be built automatically.
      Just remove the space in the following string after the + sign: "+DOCS"
- [x] If any network modules should be tested for the PR, add them as a comma separated list
      (e.g. `clients.fdsn,clients.arclink`) after the colon in the following magic string: "+TESTS:"
      (you can also add "ALL" to just simply run all tests across all modules)
- [x] All tests still pass.
- [x] Any new features or fixed regressions are be covered via new tests.
- [x] Any new or changed features have are fully documented.
- [not sure who does this?] Significant changes have been added to `CHANGELOG.txt` .
- [x] First time contributors have added your name to `CONTRIBUTORS.txt` .

### CODE REVIEW

File: obspy/signal/spectral_estimation.py
Comment (heavelock) [inline obspy/signal/spectral_estimation.py:983]: I would rather go for `if self.special_handling in ("hydrophone", "infrasound")`

File: obspy/signal/spectral_estimation.py
Comment (heavelock) [inline obspy/signal/spectral_estimation.py:2215]: This should be inside of the docstring, not as a separate comment

File: obspy/signal/spectral_estimation.py
Comment (heavelock) [inline obspy/signal/spectral_estimation.py:2228]: The same as before.

File: obspy/signal/spectral_estimation.py
Comment (heavelock) [inline obspy/signal/spectral_estimation.py:2210]: I don't think that there was a test for that method done before, could you add it?

File: obspy/signal/spectral_estimation.py
Comment (heavelock) [inline obspy/signal/spectral_estimation.py:2216]: Here, I think, the trailing underscore after citation is crucial for the Sphinx to properly interpret it as a link. Additionally, could you add a bib entry in the `obspy/misc/docs/source/bibliography` for that paper?

**Review by heavelock — CHANGES_REQUESTED** [review]
This 33MiB sac file is definitely no-go. Could you either slice it drastically down to <400KiB or even prepare something synthetic inside of the test itself?  There is a `Stream._dummy_stream_from_string` method that allows for super easy creation of synthetic streams for testing.

File: obspy/signal/spectral_estimation.py
Comment (megies) [inline obspy/signal/spectral_estimation.py:1967]: This could be a bit confusing, since `nhnm` stands for "New High Noise Model" which is a term coined in that Peterson paper. Also having this as an `arg` will break peoples' codes, potentially, or rather likely.
I'd prefer a new helper method to fetch the infrasound models instead.

```

        if self.special_handling == "infrasound":
            # Re-name noise model file
            models = (get_boide_high_noise(), get_boide_low_noise())  # or some other appropriate naming
        else:
            models = (get_nhnm(), get_nlnm())

        if show_noise_models:
            for periods, noise_model in models:
```

File: obspy/signal/tests/test_spectral_estimation.py
Comment (megies) [inline obspy/signal/tests/test_spectral_estimation.py:568]: This test is only of limited benefit right now. It basically just assures that something is done and no error pops up. There's two ways here. Either set up an image test, but that should really be avoided when possible, as it's a huge maintenance burden. So, ideally what you want to do is simply inspect the artists in the figure and do some assertions on those.

See the `test_earthquake_models` test case in `obspy/signal/tests/test_spectral_estimation.py` for an example:

https://github.com/obspy/obspy/blob/a8750f46132d937daec3a41ed6d5f38478a4ecc0/obspy/signal/tests/test_spectral_estimation.py#L544-L566

Also what @heavelock said, the test data should be at most the length such that 2 or 3 PSD pieces get processed. Adding more data does not help the test in any way and just burdens our repository (and git history) and slows down the testing.

(eventually that 30MB file needs to get rebased out of the PR history, so when you `git remove` that file, best do it in a single commit that can then easily be squashed on top of the commit that added the file in the first place)

**Review by megies — CHANGES_REQUESTED** [review]
Some commentd for changes, also the the PR file diff for PEP8 fails that need fixing.

File: obspy/signal/spectral_estimation.py
Comment (kamacpherson) [reply obspy/signal/spectral_estimation.py:983]: That is better - thanks!

File: obspy/signal/spectral_estimation.py
Comment (kamacpherson) [reply obspy/signal/spectral_estimation.py:1967]: This is much better - thank you!

File: obspy/signal/spectral_estimation.py
Comment (kamacpherson) [reply obspy/signal/spectral_estimation.py]: I think this is resolved after a created new definitions to read the infrasound models per megies's advice.

File: obspy/signal/spectral_estimation.py
Comment (kamacpherson) [reply obspy/signal/spectral_estimation.py]: I think this is resolved after a created new definitions to read the infrasound models per megies's advice.

File: obspy/signal/spectral_estimation.py
Comment (kamacpherson) [reply obspy/signal/spectral_estimation.py]: I removed this argument per megies's advice. Also working-on a new test.

File: obspy/signal/spectral_estimation.py
Comment (kamacpherson) [reply obspy/signal/spectral_estimation.py]: Thanks - I added a smaller test file, added the underscores, and added the reference to the bibliography directory.

File: obspy/signal/tests/test_spectral_estimation.py
Comment (kamacpherson) [reply obspy/signal/tests/test_spectral_estimation.py:568]: Thank you for the advice. I have been attempting to implement it, but I have managed to mess something up in the test, and I'm a little stuck. When I run the obspy-runtests, I get
`>>> Cannot import test suite for module obspy.signal due to:
------------------------------------------------------------
  File "/Users/ken/opt/anaconda3/envs/obspy_dev/bin/obspy-runtests", line 33, in <module>
    sys.exit(load_entry_point('obspy', 'console_scripts', 'obspy-runtests')())
  File "/Users/ken/obspy/obspy/scripts/runtests.py", line 744, in main
    errors = run(argv, interactive)
  File "/Users/ken/obspy/obspy/scripts/runtests.py", line 712, in run
    pr_url=args.pr_url)
  File "/Users/ken/obspy/obspy/scripts/runtests.py", line 537, in run_tests
    suites, status, import_failures = _get_suites(verbosity, tests)
  File "/Users/ken/obspy/obspy/scripts/runtests.py", line 161, in _get_suites
    suite.append(ut.loadTestsFromName(test, None))
  File "/Users/ken/opt/anaconda3/envs/obspy_dev/lib/python3.6/unittest/loader.py", line 204, in loadTestsFromName
    test = obj()
  File "/Users/ken/obspy/obspy/signal/tests/__init__.py", line 12, in suite
    add_doctests(suite, MODULE_NAME)
  File "/Users/ken/obspy/obspy/core/util/testing.py", line 115, in add_doctests
    fromlist=["obspy"])

ModuleNotFoundError: No module named 'obspy.signal.'`

Comment (kamacpherson) [reply obspy/signal/tests/test_spectral_estimation.py:570]: I don't know what changed, but the test runs fine now.

**Comment by kamacpherson** [comment]
Thanks for your help. I believe I have addressed the requested changes.

File: obspy/signal/spectral_estimation.py
Comment (heavelock) [inline obspy/signal/spectral_estimation.py:404]: Why that deletion?

**Comment by heavelock** [comment]
I have just one more question and I think it's ready to be merged

File: obspy/signal/spectral_estimation.py
Comment (kamacpherson) [reply obspy/signal/spectral_estimation.py:404]: Yikes - this was an inadvertent deletion. Thanks for catching that. I will add it back in.

Comment (heavelock) [reply obspy/signal/spectral_estimation.py:404]: Thanks!

**Comment by heavelock** [comment]
Oh, one more thing! Could you add a changelog entry?

**Comment by megies** [comment]
Before merging we'll want to manually squash c07ecc4 on top of the commit that introduced those files to get rid of them from history and not blow up the git repo. Let me know if you need clearer instructions or need me to do it.

File: CHANGELOG.txt
Comment (megies) [inline CHANGELOG.txt:7]: Please keep the changelog in alphabetical order. The only exception we usually make is to have `obspy.core` at the very top, but then everything else beneath in alphabetical order. I see another breach with the FDSN client log at the bottom, maybe while you're at it you can move that in the right spot too?

File: obspy/signal/spectral_estimation.py
Comment (megies) [inline obspy/signal/spectral_estimation.py:54]: just a comment, since master is now Python >=3.6 we could use pathlib for stuff like that getting added which is less ugly

```python
Path(__file__).parent / "data" / "idc_noise_models.npz"
```

File: CHANGELOG.txt
Comment (kamacpherson) [reply CHANGELOG.txt:7]: Thanks - will do!

**Comment by megies** [comment]
I think this is ready to merge, after rebasing to get rid of the merge commit and squashing the big files out of the history. @kamacpherson let me know if you need help or want me to do it.

**Comment by heavelock** [comment]
I agree that's it's ready to merge. 
Only one tiny thing, in the changelog you forgot to remove one of the conflict indicators: `>>>>>>> master`

**Review by heavelock — APPROVED** [review]

**Comment by kamacpherson** [comment]
> I think this is ready to merge, after rebasing to get rid of the merge commit and squashing the big files out of the history. @kamacpherson let me know if you need help or want me to do it.

@megies I don't know how to do this to be honest:( Would you mind either giving me some guidance or just doing it? Whichever is less painful for you!

**Comment by megies** [comment]
> Would you mind either giving me some guidance or just doing it? Whichever is less painful for you!

I can do both, actually. :wink: 

How are you working with git? Command line or some GUI?

**Comment by kamacpherson** [comment]
> > Would you mind either giving me some guidance or just doing it? Whichever is less painful for you!
> 
> I can do both, actually. 😉
> 
> How are you working with git? Command line or some GUI?

Thank you @megies! I am just using the command line.

**Comment by megies** [comment]
@kamacpherson  Make sure you have a clean local state (no local changes, stash them if needed). Then on command line you'd do something like

```bash
# git remote -vv  # if you need to check aliases
git fetch origin  # or what alias you have for the obspy main repo
git rebase -i origin/master
```

Editor of your choice should open, now move the commit where you removed those big files directly under the commit that introduced them and change the command on that line to "f" or "fixup". Save the text file and should be done, `git push -f`. Can't really make errors, can always go back to the old state. Nowadays github even shows the old commit hash on force pushes, so its even easier to go back.

File: CHANGELOG.txt
Comment (megies) [inline CHANGELOG.txt:10]: This ended up in the middle of the `obspy.core` block, somehow. Please move it down.

**Comment by megies** [comment]
Somehow there is still a merge commit in here.. that shouldn't be here if you rebase on current master branch if shown above. If you can't get rid of it I can fix it

**Comment by WayneCrawford** [comment]
Here's a silly and too-late question: why does the "hydrophone" special handling plot as "dB" (no units)?  I think their responses are always specified as Pascals, like infrasound sensors.

Cheers
Wayne Crawford

**Comment by megies** [comment]
I believe units in the plot are adjusted accordingly, @WayneCrawford?

https://github.com/obspy/obspy/pull/2740/files#diff-e8c8ca46751850ef25283979a109ab1e0c0472316fc52979cdfdca42735cde8dR1650

**Comment by kamacpherson** [comment]
> Somehow there is still a merge commit in here.. that shouldn't be here if you rebase on current master branch if shown above. If you can't get rid of it I can fix it

I'm struggling a bit with this; would you mind fixing it @megies?

**Comment by megies** [comment]
@kamacpherson I can't force push to your branch. There should be a box to tick on the right "allow edits to maintainers" or something..

**Comment by kamacpherson** [comment]
> @kamacpherson I can't force push to your branch. There should be a box to tick on the right "allow edits to maintainers" or something..

@megies Sorry if I'm becoming a nuisance but I can't seem to find this button anywhere. Should it be on the pull request page?

**Comment by megies** [comment]
@kamacpherson github docs should help: https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

That branch seems to be on an organizational account? Not sure if that has any implications.

**Comment by kamacpherson** [comment]
> @kamacpherson github docs should help: https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork
> 
> That branch seems to be on an organizational account? Not sure if that has any implications.

Ah, it was the org account. I transferred the fork to me and the check box appeared!

**Comment by megies** [comment]
> I transferred the fork to me and the check box appeared!

Oh, I did not even know you could transfer a PR from one fork to another, interesting..
In any case, I force pushed the branch rebase just now.

**Comment by megies** [comment]
I noticed that actually the computed PSD data wasn't tested @kamacpherson, so I added testing those, and I also further shortened the test data to 14kb. No need to have long time series for this test.

**Review by megies — APPROVED** [review]

**Comment by jdassink** [comment]
Thanks mate for adding this. Now I finally throw out those ObsPy tweaks in my scripts..

I haven't looked carefully at the code,  but can the user select the unit for the PPSDs? Or is it only dB re 20e-6 Pa?

**Comment by megies** [comment]
> Or is it only dB re 20e-6 Pa?

Not sure I understand what you're saying.

**Comment by jdassink** [comment]
Typically the Obspy PPSD plots are plotted in dB referenced to some value. Of course this value can be chosen arbitrarily. In the case of seismic data, the PPSDs are plotted as acceleration with a reference value taken to be 1 m/s**2. However, in the study of atmospheric acoustics, some people tend to take 20e-6 Pa as reference value (hence the dB re 20e-6 Pa). Ironically, this value has little meaning in the case of infrasound; in the audible band it corresponds to the lowest audible sound level. In the ocean this is 1e-6 Pa.

What I mean to say, is that I think it would be useful that people:

1. Could select if they want dB or the actual unit (so in the case of Pa) on the ordinate.
2. Can specify which reference value they want to take in the logarithm. Default can be one.

**Comment by megies** [comment]
> dB re 20e-6 Pa?

Well yes, I was just confused because I don't see that being done in this PR, but maybe I'm overlooking something.

In any case, I'm not working with infrasound data, so you guys figure out what's needed/wanted.

**Comment by WayneCrawford** [comment]
I would be for the option suggested by Jelle.  It would be useful for hydrophone data as well.  It could be a separate PR if that works better.

Regards
Wayne

_______________________________
Wayne Crawford
CNRS Researcher
Institut de Physique du Globe de Paris
1, rue Jussieu
75238 Paris Cedex 05
+33 6 5151 1054

On 9 December 2020 at 10:38:31, Tobias Megies (notifications@github.com) wrote:

dB re 20e-6 Pa?

Well yes, I was just confused because I don't see that being done in this PR, but maybe I'm overlooking something.

In any case, I'm not working with infrasound data, so you guys figure out what's needed/wanted.

**Comment by kamacpherson** [comment]
> I would be for the option suggested by Jelle.  It would be useful for hydrophone data as well.  It could be a separate PR if that works better. Regards Wayne
> […](#)
> _______________________________ Wayne Crawford CNRS Researcher Institut de Physique du Globe de Paris 1, rue Jussieu 75238 Paris Cedex 05 +33 6 5151 1054 On 9 December 2020 at 10:38:31, Tobias Megies (notifications@github.com) wrote: dB re 20e-6 Pa? Well yes, I was just confused because I don't see that being done in this PR, but maybe I'm overlooking something. In any case, I'm not working with infrasound data, so you guys figure out what's needed/wanted. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

For this PR I assumed the reference pressure is 1 Pa, but it would definitely be nice to make that user-specified. I think adding that capability should be a new PR?

**Comment by megies** [comment]
> For this PR I assumed the reference pressure is 1 Pa, but it would definitely be nice to make that user-specified. I think adding that capability should be a new PR?

Yea, just make a new PR.

**Pull request merged**
